pkg/controller/apicast/apicast_logic_reconciler.go

```diff
@@ -0,0 +1,459 @@
+	}
+}
+
+func (r *APIcastLogicReconciler) addOwnerRef(obj metav1.Object) error {
+	ro, ok := obj.(runtime.Object)
+	if !ok {
+		return fmt.Errorf("is not a %T a runtime.Object, cannot call addOwnerRef", obj)
+	}
+	err := controllerutil.SetControllerReference(r.APIcastCR, obj, r.Scheme())
+
+	if err != nil {
+		r.Logger().Error(err, "Error setting OwnerReference on object. Requeuing request...",
+			"Kind", ro.GetObjectKind(),
+			"Namespace", obj.GetNamespace(),
+			"Name", obj.GetName(),
+		)
+	}
+	return err
+}
+
+func (r *APIcastLogicReconciler) reconcileAdditionalEnvSecret() error {
+	if r.APIcastCR.Spec.EnvironmentConfigurationSecretRef.Name == "" {
+		return fmt.Errorf("Field 'name' not specified for Additional Env Secret Reference")
+	}
+
+	environmentConfigurationSecret := v1.Secret{}
+	err := r.Client().Get(context.TODO(), types.NamespacedName{Name: r.APIcastCR.Spec.EnvironmentConfigurationSecretRef.Name, Namespace: r.APIcastCR.Namespace}, &environmentConfigurationSecret)
+	if err != nil {
+		return err
+	}
+
+	r.addOwnerRef(&environmentConfigurationSecret)
+	if err != nil {
+		return err
+	}
+
+	err = r.Client().Update(context.TODO(), &environmentConfigurationSecret)
+	return err
+}
+
+func (r *APIcastLogicReconciler) reconcileDeployment(desiredDeployment appsv1.Deployment, adminPortalSecret *v1.Secret) error {
+	adminPortalConfigurationSecret := v1.Secret{}
+	err := r.Client().Get(context.TODO(), types.NamespacedName{Name: adminPortalSecret.Name, Namespace: desiredDeployment.Namespace}, &adminPortalConfigurationSecret)
+	if err != nil {
+		return err
+	}
+
+	existingDeployment := appsv1.Deployment{}
+	err = r.Client().Get(context.TODO(), r.namespacedName(&desiredDeployment), &existingDeployment)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			desiredDeployment.Spec.Template.Annotations[AdmPortalSecretResverAnnotation] = string(adminPortalConfigurationSecret.ResourceVersion)
+			err = r.Client().Create(context.TODO(), &desiredDeployment)
+			r.Logger().Info("Creating Deployment...")
+			return err
+		}
+		return err
+	}
+
+	changed := false
+
+	if !reflect.DeepEqual(existingDeployment.Spec.Replicas, desiredDeployment.Spec.Replicas) {
+		existingDeployment.Spec.Replicas = desiredDeployment.Spec.Replicas
+		changed = true
+	}
+	if !reflect.DeepEqual(existingDeployment.Spec.Template.Spec.Containers[0].Image, desiredDeployment.Spec.Template.Spec.Containers[0].Image) {
+		existingDeployment.Spec.Template.Spec.Containers[0].Image = desiredDeployment.Spec.Template.Spec.Containers[0].Image
+		changed = true
+
+	}
+	if !reflect.DeepEqual(existingDeployment.Spec.Template.Spec.ServiceAccountName, desiredDeployment.Spec.Template.Spec.ServiceAccountName) {
+		changed = true
+		existingDeployment.Spec.Template.Spec.ServiceAccountName = desiredDeployment.Spec.Template.Spec.ServiceAccountName
+
+	}
+
+	if !reflect.DeepEqual(existingDeployment.Spec.Template.Spec.Containers[0].Env, desiredDeployment.Spec.Template.Spec.Containers[0].Env) {
```

Comment:
Not sure whether we should first sort both arrays before performing the comparison

Reply:
I think you should. Maybe not in this PR. 

Or, reconcile one by one. Each approach has trade-offs.